### ISSUE TITLE
Provide informative message for bad chunking values

### ISSUE THREAD

**Issue by tclune** [GEOS-ESM/MAPL#626]
If the requested chunking in netcdf is inconsistent with the grid dimensions, an informative error message should be provided.   This might already be the case, but recent work suggests that the failure is subtle.

**Comment by bena-nasa** [comment]
This is a more pervasive problem that just chunking. In general errors on the server side are not being trapped because of inconsistent use of return codes and the concurrent nature. For example, when you do pass a bad chunking the call does throw the error as Bill is seeing but does not stop the code. I did look at this tried to chase it up the chain. At some point the OO becomes difficult to trace and there are a lot of paths to check so this is not just a quick mornings work.

**Comment by bena-nasa** [comment]
Here is what I started trying to trap an error on the variable create. I got to the point where I could get the code hang but not cleanly crash when running a concurrent server.
https://github.com/GEOS-ESM/MAPL/tree/feature/bmauer/pfioserver_errorhandling

**Comment by weiyuan-jiang** [comment]
@mathomp4 , I don't think that is intended. When we developed the pFIO, we didn't have a 'standard' way for the  return codes.  I added that latter on and obviously missed a lot of places

**Comment by tclune** [comment]
Perhaps the 1st step is just to do a code review (min 2 people) where the focus is entirely on error handling.   E.g.,  look at each procedure interface and ensure that it has an RC dummy argument (unless it can be proven to not need one).  And ensure that each procedure uses the `_RETURN` macro (again unless it does not need one).

That should also prime the participants for thinking about how to improve the ability to propagate failures back to the client side.

**Comment by bena-nasa** [comment]
Looks like I was close in the branch, almost to the top of the server. I just pushed my branch to demonstrate that an MPI_Abort in the MPIServer is able to trap and kill on error in the file creation for reference. Of course we have the simple server, the new mutli-layered servers, other error possibilities.

**Comment by bena-nasa** [comment]
Pushed again, got it to crash when running the shared (Simple) server on the file create.  Lots of places in ClientThread.F90 need return status checks added.

**Comment by mathomp4** [comment]
@bena-nasa @weiyuan-jiang  Have you essentially solved this by now? I think our returns are a lot better now.

**Comment by mathomp4** [comment]
So I decided to see if this was still an issue and I guess I am dumb. I tried adding a chunksize line to `HISTORY.rc` in a C90 run:
```
...
  geosgcm_prog.grid_label: PC360x181-DC
  geosgcm_prog.chunksize:  127 29 1 1,
...
```
thinking, hey, prime numbers aren't good chunks but apparently netCDF or MAPL is better than I know? Or I don't know what bad chunking values are...

**Comment by bena-nasa** [comment]
A bad chunking value would be one that say is greater than the size of the dimension, try that. The values you tried above are perfectly valid, I don't think there is anything that says the chunking must divide evenly.

**Comment by mathomp4** [comment]
> A bad chunking value would be one that say is greater than the size of the dimension, try that. The values you tried above are perfectly valid, I don't think there is anything that says the chunking must divide evenly.

Hmm. Well, I just tried:
```
  geosgcm_prog.grid_label: PC360x181-DC
  geosgcm_prog.chunksize:  500 200 1 1,
```
and it worked. I even tried adding:
```
  geosgcm_prog.deflate:    1,
```
thinking maybe chunking goes hand-in-hand with compression. History still worked. Maybe this doesn't matter anymore?

**Comment by bena-nasa** [comment]
hmm, does it actually respect our chunking values? I looked in History and newCFIO, I don't see any sort of protection incase the suer specifies chunking values that are too big. Maybe Weiyan implemented something on the pFIO side?

**Comment by mathomp4** [comment]
I'm going to re-open this as I'm not sure if @bena-nasa or @weiyuan-jiang have looked at how chunking is handled.

**Comment by mathomp4** [comment]
I'm long-terming this. I think @bena-nasa can probably fix this when he has time. 

For now though at least we *crash* bad chunking at write time. @bena-nasa can make it so that we crash *gracefully* at initialization before we get to that. 😄
